packages/calypso-e2e/src/lib/components/likes-component.ts

```diff
@@ -0,0 +1,80 @@
+/**
+ * Internal dependencies
+ */
+import { BaseContainer } from '../base-container';
+
+/**
+ * Type dependencies
+ */
+import { Frame, Page, ElementHandle } from 'playwright';
+
+const selectors = {
+	// Note the variation of 'Like' button for when the button is already clicked.
+	likeWidget: 'iframe.post-likes-widget',
+
+	// Selectors within the Like Widget iframe.
+	likeButton: 'text=Like',
+	likedButton: 'text=Liked',
+	likedText: 'text="You like this."',
+};
+
+/**
+ * Component representing the like component on a post/page.
+ *
+ * @augments {BaseContainer}
+ */
+export class LikesComponent extends BaseContainer {
+	frame!: Frame;
+
+	/**
+	 * Constructs an instance of the object.
+	 *
+	 * @param {Page} page Instance of the page on which the component resides.
+	 */
+	constructor( page: Page ) {
+		super( page, selectors.likeWidget );
+	}
+
+	/**
+	 * Overrides the parent method for post-initialization steps.
+	 * This ensures the iframe is enabled and interactable.
+	 *
+	 * @returns {Promise<void>} No return value.
+	 */
+	async _postInit(): Promise< void > {
+		const handle = ( await this.page.$( selectors.likeWidget ) ) as ElementHandle;
+		this.frame = ( await handle.contentFrame() ) as Frame;
+		await this.page.waitForLoadState( 'networkidle' );
+	}
+
+	/**
+	 * Clicks the Like button and toggles the state.
+	 *
+	 * This function will also confirm that click action on the Like button
+	 * had the intended effect.
+	 *
+	 * @returns {Promise<void>} No return value.
+	 */
+	async clickLike(): Promise< void > {
+		const isLiked = await this.frame.isVisible( selectors.likedText );
+
+		// Build the list of promises to be resolved at end.
+		const promises: any = [
+			// Clicking the Like triggers an navigation event.
+			// Wait for this event to complete, as the response typically takes
+			// ~1.5s.
+			this.frame.waitForLoadState( 'networkidle' ),
+		];
+
+		if ( isLiked ) {
+			// Like button is already clicked. Check for unlike status at end.
+			promises.push( this.frame.waitForSelector( selectors.likeButton ) );
+			promises.push( this.frame.click( selectors.likedButton ) );
+		} else {
+			promises.push( this.frame.waitForSelector( selectors.likedButton ) );
+			promises.push( this.frame.click( selectors.likeButton ) );
+		}
+
+		await Promise.all( promises );
```

Comment:
This approach feels unintuitive to me as we need to declare events in reverse order. What are the benefits of calling `Promise.all` vs just `await` for individual steps?

Reply:
Yeah, this is one of the grips I have with Playwright - sometimes there is a need/suggestion to use `Promise.all`, and it isn't absolutely clear to someone relatively new to JS/TS (like myself) when one should use the `Promise.all` construct and when to use sequential `await` statements.

From what I have gleaned from countless hours on the documentation:

[Navigations](https://playwright.dev/docs/navigations) that may occur asynchronously or multiple times often require the use of `Promise.all` to prevent race condition between waiting for navigation and clicking on an element.

Popups often require the use of `Promise.all` to ensure the said popup is loaded to an acceptable state, though in a follow-up PR (yet to be posted) I have implemented popup handling without using `Promise.all`.

Ultimately, I think the use of `Promise.all` helps ensure that all required actions have been processed (in _some_ order) before proceeding. I think of it as a way to safeguard uncertain parts of the code, like clicking on the `Like` button that requires some delay after clicking for the request/response to be processed then reflected on the DOM.

With the (long wall of text) said, I can try to remove the need to use `Promise.all` here. You will likely see a lot more of `Promise.all`s coming in subsequent patches though, as they are really useful in Playwright.

Reply:
Well well well, guess sequential `awaits` can work, and seems reliable too. So much for all that headache trying to make the Promise.all work reliably.

Reply:
Thanks for the context, it explains a lot! I think using `Promise.all` makes sense in the navigation/popup scenario because we need to be already in the state of waiting for the **change event** before clicking the triggering link. In case of i.e. waiting for some element to appear on the current page as a result of **state change** (i.e. 'Like' button click), it should not be necessary to use `Promise.all`. In such a case, we're only checking if an element is present as a way of self-testing the helper, which should be fine as a simple sequential `await` call. Actually, I have a feeling that using `Promise.all` might be risky in some situations as it runs all the promises in parallel. It does fire them in given sequence, but it doesn't wait for resolution before calling the next one. See this example bin: https://jsbin.com/ropuyoj/1/edit?js,console